### DISCUSSION TITLE
3 devices (7 digit max7219) cascaded with pico w

### DISCUSSION THREAD

**Discussion by MagisterTechnicus** [micropython/micropython#14052] [Hardware & Peripherals] [answered]
![image0](https://github.com/micropython/micropython/assets/58939194/b604dc8d-ec3a-485f-9ab4-34608d262934)

Is there a framework for micropythen to address 3 seperate 8 x 7 segment devices? I read, that the devices can also be controlled separately via an SPI bus, but all the attempts I made didn't work. I'm grateful for every tip

**Comment by robert-hh** [comment]
The picture shows an CS input at the devices. So it looks like you can use a common CLK and DI pin for all of them and have one separate CS pin for each.

**Comment by MagisterTechnicus** [comment]
I use this driver:
https://github.com/pdwerryhouse/max7219_8digit

```py
#Connections:
#pico w -> 7 digit device
#
#first device
#GP1, SPI0 CSn -> CS 
#GP2, SPI0 SCK -> CLK
#GP3, SPI0 TX  -> DIN
#
#second device
#GP5, SPI0 CSn -> CS 
#GP2, SPI0 SCK -> CLK
#GP3, SPI0 TX  -> DIN

from machine import Pin, SPI
import max7219_8digit
import time

spi0 = SPI(0, baudrate=100000, polarity=1, phase=0, sck=Pin(2), mosi=Pin(3))

ss1 = Pin(1, Pin.OUT)
ss2 = Pin(5, Pin.OUT)

count = 0

line1 = max7219_8digit.Display(spi0, ss1)
line2 = max7219_8digit.Display(spi0, ss2)

while True:
    line1.write_to_buffer(str(count))
    line1.display()
    
    line2.write_to_buffer(str(count * 2))
    line2.display()
       
    count = count + 1
    
    time.sleep(0.5)`
```
And this is the code. But it doesn't run properly. Can you give me an advice? Maybe I did something wrong with the wiring stuff. In the first lines you will find the assignment of the pins.

After some correct loops, the second display shows wrong data. But it's not always at the same point when counting up. Display 1 continues to work without any problems!

**Comment by steveh127** [comment]
Two things. 

Please see forum guidlines to properly comment code. Basically just surround it with triple quotes.

Even though the displays don't use the MISO line you should specify it for SPI to work. It might also help if you showed your errors rather than just stating that it doesn't work.

**Comment by mathiasflick** [comment]
Looking at your set-up (the picture) I would not completely exclude connection problems (please don't get me wrong, but your soldering looks horrible ...).
So, probably after some resoldering, I would reduce the code to drive a single display (e.g. in the upper position in the picture) and test all your display modules regarding proper functioning in this position, thus excluding connection problems (like bad soldering) and malfunction of the respective modules. Then repeat the tests in the other position (lower position, same SPI, different CS) to further exclude wiring problems. 
After those tests have passed, move on (respectively back) to the multi-display set-up. 
Good luck and greetings from Cologne

**Comment by MagisterTechnicus** [comment, marked as answer]
Here is an example that work with two (and more) LED segments in a cascade way. Simply adjust the variable digit= with the correct number (8,16,24,....). Wiring see picture two. 

hardware pin(2), GP1, SPI0 CSn -> CS 
hardware pin(4), GP2, SPI0 SCK -> CLK
hardware pin(5), GP3, SPI0 TX  -> DIN

```py
import max7219
import time

display = max7219.SevenSegment(digits=16, scan_digits=8, cs=1, spi_bus=0, reverse=True)

display.clear()
display.text("ABCDEFGI12345678")

time.sleep(10)

display.clear()
display.text("22.33           ")

time.sleep(2)

display.clear()
display.message("Hello World")

time.sleep(2)

display.clear()
```

![IMG_7539](https://github.com/micropython/micropython/assets/58939194/09aa6430-500d-4a59-8f70-d17e33dd54a0)
![IMG_7540](https://github.com/micropython/micropython/assets/58939194/de9d9df2-7256-416d-ab29-0ce024fc3d75)
